org-roam-org-ref.el

```diff
@@ -0,0 +1,281 @@
+;;; org-roam-org-ref.el --- Roam Research replica with Org-mode -*- coding: utf-8; lexical-binding: t -*-
+
+;; Copyright © 2020 Jethro Kuan <jethrokuan95@gmail.com>
+;; Copyright © 2020 Mykhailo Shevchuk <mail@mshevchuk.com>
+
+;; Author: Jethro Kuan <jethrokuan95@gmail.com>
+;; Contributor: Mykhailo Shevchuk <mail@mshevchuk.com>
+;; URL: https://github.com/jethrokuan/org-roam
+;; Keywords: org-mode, roam, convenience
+;; Version: 1.0.0-rc1
+;; Package-Requires: ((emacs "26.1") (dash "2.13") (f "0.17.2") (s "1.12.0") (org "9.3") (emacsql "3.0.0") (emacsql-sqlite "1.0.0"))
+
+;; This file is NOT part of GNU Emacs.
+
+;; This program is free software; you can redistribute it and/or modify
+;; it under the terms of the GNU General Public License as published by
+;; the Free Software Foundation; either version 3, or (at your option)
+;; any later version.
+;;
+;; This program is distributed in the hope that it will be useful,
+;; but WITHOUT ANY WARRANTY; without even the implied warranty of
+;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+;; GNU General Public License for more details.
+;;
+;; You should have received a copy of the GNU General Public License
+;; along with GNU Emacs; see the file COPYING.  If not, write to the
+;; Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+;; Boston, MA 02110-1301, USA.
+
+;;; Commentary:
+;;
+;; This library offers a tighter integration between org-roam and
+;; org-ref by providing `org-roam-org-ref-notes-fn' function,
+;; which upon installing into `org-ref-notes-function' will enable
+;; org-ref to use org-roam as a backend for managing bibliography
+;; notes.
+;;
+;; Use it like this:
+;;
+;; (setq org-ref-notes-function #'org-roam-org-ref-notes-fn
+;;       org-ref-notes-directory "path/to/org-roam-directory")
+;;
+;; Optionally, `org-roam-capture-templates' can be dynamically
+;; preformatted with bibtex field values. See
+;; `org-roam-org-ref-preformat-keywords' for more details.
+;;
+;; Optionally, automatic switching to the perspective (persp-mode)
+;; with the notes project (projectile) is possible. See
+;; `org-roam-org-ref-edit-notes' for more details.
+;;
+
+;;; Code:
+;;;; Library Requires
+(require 'org-roam)
+
+(declare-function org-ref-find-bibliography "org-ref-core")
+(declare-function projectile-relevant-open-projects "projectile")
+(declare-function persp-switch "persp-mode" (name &optional frame (window (selected-window)) (called-interactively-p (called-interactively-p 'any))))
+(declare-function persp-names "persp-mode" (&optional (phash *persp-hash*) (reverse t)))
+
+(defvar org-roam-org-ref-preformat-templates nil
+  "Non-nil to enable template preformatting.
+See `org-roam-org-ref-edit-notes' for details.")
+
+(defvar org-roam-org-ref-preformat-keywords '("citekey" . "=key=")
+  "The template prompt wildcards for preformatting.
```

Comment:
Should these be `defcustom`s rather than `defvar`? If I understand correctly it is intended for these to be user customisable.

Reply:
Thank you. Yes definitely, all the user options must be `defcustom`.